runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp

```diff
@@ -139,6 +139,82 @@ class RecognizedCallTransformer : public OMR::RecognizedCallTransformer
     *     \endcode
     */
    void process_java_lang_StringUTF16_toBytes(TR::TreeTop* treetop, TR::Node* node);
+   /** \brief
+    *     Transforms jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I
+    *     into an arraycmp, bit manipulation and iselect sequence with equivalent semantics.
+    *
+    *  \param treetop
+    *     The treetop which anchors the call node.
+    *
+    *  \param node
+    *     The call node representing a call to jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I
+    *     which has the following shape:
+    *
+    *     \code
+    *     icall  <jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I>
+    *       <a>
+    *       <aOffset>
+    *       <b>
+    *       <bOffset>
+    *       <length>
+    *       <log2ArrayIndexScale>
+    *     \endcode
+    *
+    *  \details
+    *     The call node is transformed to the following shape:
+    *
+    *     \code
+    *     iselect ()
+    *       icmpeq
+    *         arraycmp (arrayCmpLen )
+    *           aladd
+    *             <a>
+    *             <aOffset>
+    *           aladd
+    *             <b>
+    *             <bOffset>
+    *           i2l
+    *             iand
+    *               ishl
+    *                 <length>
+    *                 <log2ArrayIndexScale>
```

Comment:
I didn't want to get into making `arrayCmpLen` produce a 64 bit integer, but I've added in extending the length to avoid the length in bytes overflow issue

Reply:
@jdmpapin with respect to recognizing the method and vectorizing it, definitely a 32-bit size limit on `arrayCmpLen` would cause an issue here. You may be better familiar with usage of `arrayCmp` withing other opts like Idiom recognition. Spencer did try another unit test that in which Idiom Recognition optimizes the loop that compares two arrays and we do see that there is a test for `MaxLoopIterationGuard` generated to avoid doing arrayCmp where results could end up more than 32-bit. 

For the purpose of accelerating API we could certainly use similar logic and in slow path call the java implementation, but honestly, that would increase the path length unnecessarily. Do we even have other nodes where size is dependent on the flag you mentioned?

Reply:
I'm not aware of any opcode whose result type depends on the presence or absence of a flag, no. Really, it's already bad enough (IMO) that the meaning of the result depends on the flag. Ideally this functionality would have its own opcode, and it would produce either a 64-bit int or a pointer-sized int. (The latter means that the result type would differ depending on the size of a pointer on the target platform, but that seems much less objectionable, at least to me.)

Reply:
I agree with you, having a result pointer sized int seems reasonable to me. Having its own type, Correct me if I am wrong, but we do not have precedence for this one as well, right ? I do not think so it will be very complicated change to modify the `TR::arraycmp` node, we would have to change couple of places in the optimizer where they are used. Do you expect any other changes if we want to modify the opcode properties?

Reply:
Now that this is using `arraycmplen`, I'm marking this as resolved